File: AdaptiveBot.java
import java.util.*;

/* A more generic kind of bot, which adapts its behaviour according to the environment status.
 * It needs some previously collected knowledge about the environment (e.g. number of neutral planets in the map,
 * number of ships, etc.) and what strategy performs better in these conditions. This knowledge has to be 
 * collected beforehand (running a number of simulations of your bots in these environments) and manually coded 
 * in a data structure. Then, the DoTurn method can query it to know what strategy should be used in this turn. In 
 * this example we provide two environment variables: number of neutral planets in the map, and average growth
 * ratio of all planets in the map.
 * 
 * We provide a possible implementation using the hash adaptivityMap, which maps lists of integers (representing 
 * the environment) with names of bots. See AdaptivityMap.java
 * 
 * Complete this bot:
 * 1. Modify/extend the list of environment variables you consider interesting for adaptivity
 * 2. Collect data on how all your previous bots (BullyBot, RandomBot, HillclimbingBot, LookaheadBot and/or others) 
 * 	  perform in all maps
 * 3. Code these data in AdaptivityMap.java
 * 4. Complete the DoTurn method, adding all your previous bot implementations 
 */

public class AdaptiveBot {
	
	public static void DoTurn(PlanetWars pw) {
				
		//Retrieve environment status
		int neutralPlanets = pw.NeutralPlanets().size();
		int planetsSize = 0;
		for (Planet p : pw.Planets()) {
			planetsSize += p.GrowthRate();
		}
		planetsSize = planetsSize/pw.Planets().size();
			
		//Query AdaptivityMap to get which bot matches this status 
		String thisTurnBot = AdaptivityMap.get(neutralPlanets, planetsSize);
		//System.err.println("Neutral planets: " + neutralPlanets);
		//System.err.println("Planets size: " + planetsSize);
		//System.err.println("Selected bot: " + thisTurnBot);
		
		if (thisTurnBot == null) {
			System.err.println("WARNING: You have not entered bot data for this case. Using default bot");
			DoRandomBotTurn(pw);
		} else {
			if (thisTurnBot.equals("BullyBot")) {
				System.err.println("BullyBot is going to play this turn");
				DoBullyBotTurn(pw);
			} else if (thisTurnBot.equals("RandomBot")) {
				System.err.println("RandomBot is going to play this turn");
				DoRandomBotTurn(pw);
			} else {
				System.err.println("WARNING: Adaptivity map wants " + thisTurnBot +
									" to play this turn, but not implemented. Using default bot");
				DoRandomBotTurn(pw);
			}
		}
	}
	
	public static void DoBullyBotTurn(PlanetWars pw) {
		Planet source = null;
		double sourceScore = Double.MIN_VALUE;
		for (Planet myPlanet : pw.MyPlanets()) {
			if (myPlanet.NumShips() <= 1)
				continue;
			
			double score = (double) myPlanet.NumShips();
			
			if (score > sourceScore) {
				sourceScore = score;
				source = myPlanet;
			}
		}
		
		Planet dest = null;
		double destScore = Double.MAX_VALUE;
		for (Planet notMyPlanet : pw.NotMyPlanets()) {
			double score = (double) (notMyPlanet.NumShips());

			if (score < destScore) {
				destScore = score;
				dest = notMyPlanet;
			}
		}
		
		if (source != null && dest != null) {
			pw.IssueOrder(source, dest);
		}
	}
	
	public static void DoRandomBotTurn(PlanetWars pw) {

		Random random = new Random();
		Planet source = null;
		List<Planet> myPlanets = pw.MyPlanets();

		if (myPlanets.size() > 0) {
			Integer randomSource = random.nextInt(myPlanets.size());
			source = myPlanets.get(randomSource);
		}
		
		Planet dest = null;

		List<Planet> allPlanets = pw.NotMyPlanets();

		if (allPlanets.size() > 0) {
			Integer randomTarget = random.nextInt(allPlanets.size());
			dest = allPlanets.get(randomTarget);
		}

		if (source != null && dest != null) {
			pw.IssueOrder(source, dest);
		}
	}


	public static void main(String[] args) {
		String line = "";
		String message = "";
		int c;
		try {
			while ((c = System.in.read()) >= 0) {
				switch (c) {
				case '\n':
					if (line.equals("go")) {
						PlanetWars pw = new PlanetWars(message);
						DoTurn(pw);
						pw.FinishTurn();
						message = "";
					} else {
						message += line + "\n";
					}
					line = "";
					break;
				default:
					line += (char) c;
					break;
				}
			}
		} catch (Exception e) {

		}
	}
}

File: AdaptiveBot.py
#!/usr/bin/env python
#

""" AdaptiveBot - A bot which adapts its behaviour according to the environment characteristics.
 * It changes its strategy, based on the current environment (e.g. number of neutral planets in the map,
 * number of ships, etc.). Knowing which strategy to use has to be collected beforehand.
 * This requires running a number of games of your bots, and evaluate which bot performs best for a certain environment.
 * You should then add this to the data structure (in AdaptivityMap.java). 
 * The DoTurn method can then query this data structure to know what strategy should be used for this turn. 
 * This example provides two environment variables: the number of neutral planets on the map, and the average growth
 * ratio of these neutral planets.
 * 
 * We provide a possible implementation using the hash adaptivityMap, which maps lists of integers (representing 
 * the environment) with names of bots. See AdaptivityMap.java
 * 
 * Interesting questions (you can probably come up with other questions yourself as well):
 * 1. Can you modify or extend the environment variables we use? Maybe other things are interesting other than the number of neutral planets, and the average planet growth rate of these neutral planets.
 * 2. The table in AdaptivityMap.java is filled by us (randomly) with only two simple bots. But how should the table really look like? 
 * This means you should collect data on how all your previous bots (BullyBot, RandomBot, HillclimbingBot, LookaheadBot and/or others) perform in different environments
 * 3. Can you implement your other bot implementations in AdaptiveBot.java? Currently the only strategies are BullyBot ('DoBullyBotTurn') and RandomBot ('DoRandomBotTurn').
 * Implement the bot strategies you used to fill AdaptivityMap.java here as well.
"""
# Import the PlanetWars class from the PlanetWars module.
from PlanetWars import PlanetWars
#from AdaptivityMap import AdaptivityMap
# import RandomBot
# import BullyBot

def DoTurn(pw):  
  # Retrieve environment characteristics - features you can use to decide which bot to use for that specific map.
  # Are there characteristics you want to use instead, or are there more you'd like to use? Try it out!
  # In this example we will use the number of neutral planets and the average planet growth rate of neutral planets.
  num_neutral_planets = len(pw.NeutralPlanets())
  average_growth_rate = 0
  for p in pw.NeutralPlanets():
    average_growth_rate += p.GrowthRate()
  average_growth_rate = average_growth_rate / 1 + len(pw.NeutralPlanets())

  # adaptivity_map = AdaptivityMap()
  # # Use AdaptivityMap to get the bot which matches the current environment characteristics  
  # this_turn_bot = adaptivity_map.getBestBot(num_neutral_planets, 
  #                                         average_growth_rate)
    
  # if this_turn_bot is None:
  #   # There is no entry for the specified num_neutral_planets and average_growth_rate.
  #   RandomBot.DoTurn(pw)
  # elif this_turn_bot == "BullyBot":
  #   BullyBot.DoTurn(pw)
  # elif this_turn_bot == "RandomBot":
  #   RandomBot.DoTurn(pw)
  # else:
  #   # The bot in the entry is not supported yet.
  #   RandomBot.DoTurn(pw)


# Don't change from this point on. Also not necessary to understand all the details.
# Machinery that reads the status of the game and puts it into PlanetWars.
# It calls DoTurn.
def main():
  map_data = ''
  while(True):
    current_line = raw_input()
    if len(current_line) >= 2 and current_line.startswith("go"):
      pw = PlanetWars(map_data)
      DoTurn(pw)
      pw.FinishTurn()
      map_data = ''
    else:
      map_data += current_line + '\n'


if __name__ == '__main__':
  try:
    import psyco
    psyco.full()
  except ImportError:
    pass
  try:
    main()
  except KeyboardInterrupt:
    print 'ctrl-c, leaving ...'

